# R code used to analyze household microbiome data

This folder contains R scripts (generated using R version 4.3.1 in RStudio) used to analyze the data generated in the Household Microbiome project. Generic variables will be noted at the beginning of each script in a comment, as well as a short explanation of the script and why it was used.

The data in this project was single-end long read sequencing data (sequenced using Oxford Nanopore technology), which was then run through five databases to identify antibiotic resistance genes and virulence factor genes. The data was also analyzed using Kraken2 and Bracken for taxonomy. See cs_projects/ARG_VF_analysis for information on this. 

R used to create figures and run diversity analyses as well as statistical analysis on the data. 
